pkg/operator/controllers/muo/muo_controller_test.go

```diff
@@ -0,0 +1,180 @@
+package muo
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+	"strconv"
+	"testing"
+
+	"github.com/go-test/deep"
+	"github.com/golang/mock/gomock"
+	extensionsfake "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/fake"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kruntime "k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	arofake "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned/fake"
+	utillog "github.com/Azure/ARO-RP/pkg/util/log"
+	mock_dynamichelper "github.com/Azure/ARO-RP/pkg/util/mocks/dynamichelper"
+)
+
+func TestMUOReconciler(t *testing.T) {
+	tests := []struct {
+		name string
+		want ctrl.Result
+		// expected is a list of strings containing kind, namespace (if present), and name of each object, separated by a slash
+		expected        []string
+		expectedDeleted []string
+		enabled         bool
+		managed         string
+		wantErr         bool
+	}{
+		{
+			name:    "disabled",
+			enabled: false,
+			managed: "false",
+			want:    ctrl.Result{},
+		},
+
+		{
+			name:    "managed",
+			enabled: true,
+			managed: "true",
+			expected: []string{
+				"ClusterRole//managed-upgrade-operator",
+				"ConfigMap/openshift-managed-upgrade-operator/trusted-ca-bundle",
+				"ClusterRoleBinding//managed-upgrade-operator",
+				"CustomResourceDefinition//upgradeconfigs.upgrade.managed.openshift.io",
+				"Deployment/openshift-managed-upgrade-operator/managed-upgrade-operator",
+				"Namespace//openshift-managed-upgrade-operator",
+				"Role/openshift-config/muo-pullsecret-reader",
+				"Role/openshift-managed-upgrade-operator/managed-upgrade-operator",
+				"Role/openshift-managed-upgrade-operator/prometheus-k8s",
+				"Role/openshift-monitoring/muo-monitoring-reader",
+				"RoleBinding/openshift-config/muo-pullsecret-reader",
+				"RoleBinding/openshift-managed-upgrade-operator/managed-upgrade-operator",
+				"RoleBinding/openshift-managed-upgrade-operator/prometheus-k8s",
+				"RoleBinding/openshift-monitoring/muo-monitoring-reader",
+				"ServiceAccount/openshift-managed-upgrade-operator/managed-upgrade-operator",
+			},
+			want: ctrl.Result{},
+		},
+		{
+			name:    "managed=false (removal)",
+			enabled: true,
+			managed: "false",
+			expectedDeleted: []string{
+				"Deployment/openshift-managed-upgrade-operator/managed-upgrade-operator",
+			},
+			want: ctrl.Result{},
+		},
+		{
+			name:    "managed=blank (no action)",
+			enabled: true,
+			managed: "",
+			want:    ctrl.Result{},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
```

Comment:
This is really hard to follow :( I think this is partially because we combine tests of (at least) two units here:

* We test `Reconcile` and it's logic related to flags
* And we test deployer

IMO in `muo_controller_test.go` we should have a simple(ish) test (probably just one for `Reconcile`) where we have a mock of a deployer and test things like:

* What if controller is disabled
* What if `managed` is `"true"`
  * What if `deployer.CreateOrUpdate` returns `err != nil`?
  * What if `deployer.CreateOrUpdate` returns `err == nil`?
* What if `managed` is `"false"`
  * What if `deployer.Remove` returns `err != nil`?
  * What if `deployer.Remove` returns `err == nil`?
* What if `managed` is `""`
* Something else?

And we should also have `deploy_test.go`. In this one we should mock dynamic helper (maybe something else?).

I think it will be enough to check that `CreateOrUpdate` (one test):
* Creates right resource types (so we don't duplicate yamls in tests)
* Check that we replace image with the right pull spec

And check that `Remove` (a separate test, I think):

* Removes the right resource types

Everything else should be tested inside the dynamic helper and is out of scope for this.

---

Sorry for the brain dump. Does any of this make sense?

Reply:
I think I've implemented most of this, now.

Reply:
It is now a bit better in terms of readability: we no longer test deployer via this test. But it is still very hard to follow. I spent more than an hour trying to understand this test and what cases it covers.

What makes it hard in my opition:

1. A lot of flags/field in each test case with a lot of permutations. We are trying to reduce code duplication, but it hurts readability.
2. Mocking dynamic helper instead of mocking deployer

The above means that it is very time consuming to add a new test case because you need to find a right combination of test case flags. And it is hard to follow which calls are expected in which test case. 

And we need to jump between too many things to understand the test: test case struct, test body, between `Reconciler` implementation (subject of the test) and between the deployer (which should not be our concern here at all as it is tested separately) as well as `ready.CheckDeploymentIsReady` (also not our concern - it is like 2 steps down from what we are actually testing).

IMO, we should introduce an interface for deployer and mock it. I belive we will have to make it exportable to be able to generate a mock, but it is a small price to pay for readability.

I think we can reduce the test case to the following fields:

* Test case name
* A function which sets up mocks for each test case
* Probably, a function that modifies `arov1alpha1.Cluster` in a way we want before the test
* Wanted err message `string`

And the test body will be things common to al ltest cases